Mage/src/main/java/mage/watchers/Watcher.java

```diff
@@ -75,18 +77,61 @@ public void reset() {
         condition = false;
     }
 
-    protected String getBasicKey(){
+    protected String getBasicKey() {
         return getClass().getSimpleName();
     }
 
     public abstract void watch(GameEvent event, Game game);
 
-    public <T extends Watcher> T copy(){
+    public <T extends Watcher> T copy() {
         try {
-            Constructor<? extends Watcher> constructor = this.getClass().getDeclaredConstructor(getClass());
+            List<?> constructors = Arrays.asList(this.getClass().getConstructors());
+            if (constructors.size() > 1) {
+                logger.error(getClass().getSimpleName() + " has multiple constructors");
+                return null;
+            }
+
+            Constructor<? extends Watcher> constructor = (Constructor<? extends Watcher>) constructors.get(0);
+            if (constructor.getParameterCount() > 0) {
+                logger.error(getClass().getSimpleName() + " constructor has arguments, should be 0 and inject the parameters by setters");
+                return null;
+            }
             constructor.setAccessible(true);
-            return (T) constructor.newInstance(this);
-        } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+            T watcher = (T) constructor.newInstance();
+            List<Field> allFields = new ArrayList<>();
+            allFields.addAll(Arrays.asList(getClass().getDeclaredFields()));
+            allFields.addAll(Arrays.asList(getClass().getSuperclass().getDeclaredFields()));
+            for (Field field : allFields) {
+                field.setAccessible(true);
+                if (field.getType().isPrimitive()) {
```

Comment:
Having a list of supported types to copy seems hard to maintain (devs might not even know they need to add something here).  I'd just copy everything and have an easy way for developers to override.

Reply:
Correct, classes should be treated equal.

Reply:
In theory you could change the end of this if/else block from:

```
                } else {
                    if (field.getType() != Logger.class) {
                        logger.error(field.getType() + " can not be copied");
                    }
                }
```

to something like:

```
                } else if (field.getType() != Logger.class) {
                    logger.error(field.getType() + " can not be copied");
                } else {
                    field.set(watcher, field.get(this));
                }
```

Then you can delete any of the else/if clauses that don't have special handling of the type.  I think that's at least the case for deleting `isPrimitive `, `Step`, `Mana`, `UUID`,  enum, and leaving only map, set and list to handle here.

Reply:
isPrimitive is needed for , well, primitives, I will remove the custom objects, and I will investigate if Enum is explicitly needed.